Make Painless the Default Language

jdconrad opened the pull request:
As the title says.

I added the breaking label because there are places where it's possible users are using Groovy features that Painless does not have/support.  However, Painless does cover common cases where the user isn't doing something like attempting to log out of a script or other things that should probably not be done there.

dakrone commented on core/src/main/java/org/elasticsearch/script/ScriptSettings.java:
```diff
@@ -59,7 +59,7 @@ public ScriptSettings(ScriptEngineRegistry scriptEngineRegistry, ScriptContextRe
 
         this.defaultScriptLanguageSetting = new Setting<>("script.default_lang", DEFAULT_LANG, setting -> {
-            if (!"groovy".equals(setting) && !scriptEngineRegistry.getRegisteredLanguages().containsKey(setting)) {
+            if (!"painless".equals(setting) && !scriptEngineRegistry.getRegisteredLanguages().containsKey(setting)) {
```
Perhaps use `DEFAULT_LANG` here instead of hardcoding `"painless"`?

rjernst replied:
Is this first clause even necessary? I don't understand why painless (or groovy before) get's a "free pass" here from existing as a registered engine.

dakrone replied:
Good  point, that'd be even better

jdconrad replied:
Discussed this with @rjernst.  There are integ tests that depend on this check since the default lang won't be properly registered at test time.  Changing this to a constant as @dakrone suggests.

dakrone commented on modules/lang-groovy/src/test/java/org/elasticsearch/script/groovy/GroovyIndexedScriptTests.java:
```diff
@@ -136,7 +136,7 @@ public void testDisabledUpdateIndexedScriptsOnly() {
         try {
             client().prepareUpdate("test", "scriptTest", "1")
-                    .setScript(new Script("script1", ScriptService.ScriptType.STORED, GroovyScriptEngineService.NAME, null)).get();
+                    .setScript(new Script("script1", ScriptService.ScriptType.STORED, "groovy", null)).get();
```
I think you should still be able to use the constant here? In case one day we decide to rename it "GrOoVy"

jdconrad replied:
Will do -- I noticed most of the tests used "groovy" so I changed it for consistency, but I'll flip it the other way.

dakrone commented on modules/lang-groovy/src/test/java/org/elasticsearch/script/groovy/GroovyScriptTests.java:
```diff
@@ -85,8 +85,7 @@ public void testGroovyExceptionSerialization() throws Exception {
                     .setQuery(
-                            constantScoreQuery(scriptQuery(new Script("1 == not_found", ScriptType.INLINE, GroovyScriptEngineService.NAME,
-                                    null)))).get();
+                            constantScoreQuery(scriptQuery(new Script("1 == not_found", ScriptType.INLINE, "groovy", null)))).get();
```
Same here?

jdconrad replied:
Same note as above.

dakrone replied:
I think this one can still use the constant right? Maybe I'm missing which above comment you're referring to

jdconrad replied:
Fixed these... accidentally missed them the first time around.

jdconrad commented:
@dakrone and @rjernst Changed the magic values to constants.

dakrone commented:
@jdconrad I think this definitely needs a big blurb in the migration guide about the switch, with links to the documentation and all that

clintongormley commented:
This is going to break percolators and watches that use scripts but don't specify a `lang`.  Please could you mention in the migration changes that the default language can be set to back to groovy by setting `script.default_lang: groovy`

jdconrad commented:
@clintongormley This will be my first time modifying this document.  Should I only write a blurb under the migration/migrate_5_0/scripting.asciidoc specifically?  Or are there other places this needs to be mentioned as well?

jdconrad commented:
@dakrone @clintongormley I updated the migration doc to include a blurb about the switch.

dakrone commented on docs/reference/migration/migrate_5_0/scripting.asciidoc:
```diff
@@ -1,6 +1,54 @@
 [[breaking_50_scripting]]
 === Script related changes
 
+==== Switched Default Language to Painless
```
Mention (for people that might not have a lot of history with ES) that this was migrated **from** Groovy to Painless

jdconrad replied:
Done.

dakrone commented on docs/reference/migration/migrate_5_0/scripting.asciidoc:
```diff
@@ -1,6 +1,54 @@
+==== Switched Default Language to Painless
+
+The default scripting language for Elasticsearch is now Painless.  Painless is a language similar to Groovy, except
+faster and more secure.  Many Groovy scripts will be identitical to Painless scripts to help make the transition
```
"Painless is a language similar to Groovy, except faster and more secure" -> "Painless is a custom-built language with syntax similar to Groovy designed to be fast as well as secure"

jdconrad replied:
Done.

dakrone commented on docs/reference/migration/migrate_5_0/scripting.asciidoc:
```diff
@@ -1,6 +1,54 @@
+It is also possible to set the default language back to Groovy using the following setting: `script.default_lang: groovy`
+
+One common difference to note between Groovy and Painless is the user of parameters -- all parameters in Painless
+must be prefixed with `params` now.  The following example shows the difference:
```
`params` -> `params.`

jdconrad replied:
Done.

dakrone commented on docs/reference/migration/migrate_5_0/scripting.asciidoc:
```diff
@@ -1,6 +1,54 @@
+between languages as simple as possible.
+
+Documentation for Painless can be found at
+https://www.elastic.co/guide/en/elasticsearch/reference/master/modules-scripting-painless.html
```
I think there's a way to do relative links so this doesn't break in the future if we change branch names, maybe @clintongormley knows how?

clintongormley replied:
Yes, replace the absolute link with `<<modules-scripting-painless,Painless Scripting Language>>`

jdconrad replied:
Done.

clintongormley commented on docs/reference/migration/migrate_5_0/scripting.asciidoc:
```diff
@@ -1,6 +1,54 @@
+
+It is also possible to set the default language back to Groovy using the following setting: `script.default_lang: groovy`
+
+One common difference to note between Groovy and Painless is the user of parameters -- all parameters in Painless
```
user -> use

jdconrad replied:
Good catch.  Done.

clintongormley commented:
Two small comments, otherwise docs LGTM

djschny commented:
Sorry for coming late to this, but this feels premature. v5 will be the very first version of Elasticsearch that contains the new scripting language of Painless. To make it default before the community has time to test, start to adopt, etc. seems pre-mature from an external/user perspective.

Seeing this happen in a deprecation and then switch to it by default in the next major release after v5 seems much more reasonable.

What concerns me (assuming I'm reading between the lines on the implementation) is that once upgraded if a person had scripts being used that did not explicitly declare `lang` will start to execute as Painless unless they add that declaration prior to upgrading.

Since many simple groovy scripts will run fine as painless, this is a dangerous false sense that someone would be OK.

I would very much prefer this to be changed back and the switched later on as the community actually has a chance to adopt the language.

rjernst commented:
@djschny Painless was designed and developed from the start to be a replacement for groovy scripting. The syntax being essentially a subset of groovy was to ease pain of upgrading and allow the vast majority of scripts out there to "just work". Delaying the update to default painless won't help anyone, it only hurts. Either a users script will simply continue working after the upgrade (and see performance gains), or it will not compile, and they need only to specify the language explicitly (which they can already do in all versions of elasticsearch, no need to upgrade).

> What concerns me (assuming I'm reading between the lines on the implementation) is that once upgraded if a person had scripts being used that did not explicitly declare lang will start to execute as Painless unless they add that declaration prior to upgrading.

> Since many simple groovy scripts will run fine as painless, this is a dangerous false sense that someone would be OK.

This is entirely the point of the tedious work by @jdconrad and @rmuir that went into making many many groovy features and syntax available in painless. We have trappy problems like #20097 with groovy performance. Forcing most users that don't know any better (ie just use whatever the default language is) to have degraded performance isn't good. We can help most of those users by changing the default. And it's not dangerous; it is more performant and safer by design.

s1monw commented:
@djschny folks that really don't wanna take the risk can still go and set `script.default_lang: groovy` in their yaml file. it's still a module

djschny commented:
Based upon the conversation, perhaps it makes more sense to have the `lang` required whenever using inline scripts? That way there is no ambiguity.

jdconrad commented:
Updated the migration doc based on feedback and fixed some more tests.  I am for adding the lang as required, but as a separate issue.  I don't want this to be held up because of that.  If a script works it's transparent to the user, and if not the user is using more advanced features and should be able to figure out what's going on based on the error message and migration docs.  There are now several paragraphs in the migration doc on only for this change.

rmuir commented:
+1 to push. I don't see any actual technical objections, just baseless whining and resistance to change.

jdconrad commented:
This was merged -- I squashed since I had some poor commit messages on the branch.

jdconrad commented:
Thanks to all who reviewed!

s1monw commented:
> Based upon the conversation, perhaps it makes more sense to have the lang required whenever using inline scripts? That way there is no ambiguity

can you open a new issue for this, I think it's a good idea to discuss

djschny commented:
> can you open a new issue for this, I think it's a good idea to discuss

sure thing, just did on https://github.com/elastic/elasticsearch/issues/20122 and tried to add as much explanation/rationale in regards to why I had brought the topic up
